### DISCUSSION TITLE
View - Coding Conventions

### DISCUSSION THREAD

**Discussion by jeonghye-choi** [githru/githru-vscode-ext#14] [View]
### 목표

8/16일 회의를 기반으로, 함께 논의해야하는 컨벤션 항목에 대한 의견 나누기

- 폴더 구조
- Lint 설정
- 컴포넌트 선언 키워드
- 타입 선언 방법
- 머지 기준
- Material-UI 사용

### 참고사항

자유롭게 의견을 남겨주세요!

**Comment by jeonghye-choi** [comment]
### 1) 폴더 구조

회의를 통해 아래와 같이 의견이 나왔습니다.
- 페이지 이동이 없기 때문에, App.ts에서 컴포넌트 불러오기
- 컴포넌트 이름을 feature 이름으로 해주기
- 공유할 수 있는 함수에 대해서 utils 폴더로 분리하기

관련되어 다른 의견이 있으신 분은 자유롭게 reply로 코멘트 해주세요!

**Reply by jin-Pro** [reply]
Components 폴더 만들어서 feature단위로 폴더 만들고 그 하위에 feature를 또 구분하는 컴포넌트가 있으면 좋을 것 같아요!!!

Components/VerticalClusterList/index.tsx
Components/VerticalClusterList/Graph.tsx
Components/VerticalClusterList/Detail.tsx

ㅎㅎ,, 다른 의견도 들어보고싶습니다!!

**Reply by hanseul-lee** [reply]
@jin-Pro 님 의견에 동의합니다~

**Reply by pshdev1030** [reply]
저도 @jin-Pro 님 의견에 동의합니다!

**Comment by jeonghye-choi** [comment]
### 2) Lint 설정

회의를 통해 아래와 같이 의견이 나왔습니다.
- eslint, prettier는 전체 따라가기(with analysis-engine team)
- ts 부분만 정하면 됨

관련되어 다른 의견이 있으신 분은 자유롭게 reply로 코멘트 해주세요!

**Reply by jin-Pro** [reply]
engine팀이 잘해주셔서 따라가도 좋을 것 같아요!
인원이 많아 조금은 빡빡해도 좋을것 같다는 생각도 들어요..! 정 안되면 진행하면서 rule을 추가하는것도 좋을 것 같아요 !

**Reply by jejecrunch** [reply]
진행하면서 rule 추가하는 건 자주 하지는 않았으면 좋겠습니다 !

**Comment by jeonghye-choi** [comment]
### 3. 컴포넌트 선언 키워드

**functions** vs **const**

**Reply by jeonghye-choi** [reply]
(지혜님) const의 경우는 익명함수라서 function이 더 낫다는 얘기를 들었습니다.

**Reply by jeonghye-choi** [reply]
(someone..) 컴포넌트는 주로 const로 util함수는 주로 function 로 사용합니다!

**Reply by jin-Pro** [reply]
@jejecrunch  const의 경우 컴포넌트를 익명함수로 선언되어 발생하는 단점을 아실까요??? 컴포넌트가 호이스팅 이슈로 인해 발생하는 불이익은 없을 것 같아서요!

음!! ㅎㅎ,, 현업에서는 어떻게 사용하시나요???

**Reply by hanseul-lee** [reply]
저는 실무에서 거의 const로 사용하고 있습니다~!

**Reply by pshdev1030** [reply]
저는 컴포넌트로 const로 사용하고 혹시 명시적으로 바인딩 해줄 일이 있지 않을까 싶어서 유틸함수는 function으로 사용하는데요. 아직까지 한 번도 명시적으로 바인딩을 해줄 일은 없었습니다..ㅎㅎ

**Reply by jejecrunch** [reply]
@jin-Pro 님 말씀대로 생각해보면 컴포넌트라 호이스팅 이슈는 딱히 없을 것 같아요 !

**Comment by jeonghye-choi** [comment]
### 4. 타입 선언 방법

**type** vs **interface**

**Reply by jeonghye-choi** [reply]
(영택님) 이 둘의 차이는 interface를 사용하면 밑에서 한 번 더 정의했을 때 확장된다는 점을 제외하고는 type과 interface의 기능은 똑같다고 봐도 무방합니다.

**Reply by jin-Pro** [reply]
영택님께서 타입 선언을 type으로 해주셔서 type으로 그대로 진행하는 것도 좋을 것 같아요!

**Reply by hanseul-lee** [reply]
큰 차이는 없지만 공식 문서에서 interface를 권장하고 있어 interface로 통일하면 좋을 것 같습니다~

cf.
- https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/basic_type_example/#types-or-interfaces
- https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#differences-between-type-aliases-and-interfaces

**Reply by jin-Pro** [reply]
감사합니다 ㅎㅎ 읽어보겠습니다!

**Reply by jin-Pro** [reply]
공식문서에서 type을 고려하라는것 같아요!

**Reply by pshdev1030** [reply]
저도 개별 prop은 type으로 작성하더라도 interface로 통일하면 좋을 것 같습니다!

**Reply by jejecrunch** [reply]
그러면 props들의 경우는 type으로 작성하고 그외의 타입은 interface로 가는 걸까요?!

**Reply by hanseul-lee** [reply]
이부분 type으로 통일하여 작성, 이후 확장 가능성이 있다면 interface로 변경하기로 8/31 오프라인 미팅에서 결정되었습니다

**Comment by jeonghye-choi** [comment]
### 5. 머지 기준

- setState를 할 때 기본 타입을 null로 할지 undefined로 할지, empty value로 할지,… array는 undefined로 할지 [ ] 로 할지

**Reply by jin-Pro** [reply]
이거는 머지기준 타이틀하고 맞지는 않는 것 같아요!

머지 기준이 아마 승인 2명과 ci 테스트로 굳혀진것 같아요 문기님께서 script 작성 잘 해주셔서 덕분에 그대로 갈 수 있겠네요!

**Reply by jin-Pro** [reply]
array는 [] 이 좋아보입니다! 
undefined라면, 따로 처리를 해주어야 될 것 같아요.
빈배열이면 고차함수 를 사용해도 동작을 하지 않아서 좋은 것 같아요!

**Reply by jin-Pro** [reply]
그 외에서 null과 undefined는 의미적 차이가 중요할 것 같아요~
undefined는 아직 선언이 되지 않았음을 뜻하고
null은 그 자체로 값이 없다는 것을 뜻한다고 알고있어요.

이후 값이 들어오지만, 그 값이 동일한 타입을 유지한다면 undefined를 사용하는게 좋을 것 같아요!
하지만, 이후 값이 null과 섞여서 들어온다면 null을 사용하는게 좋을 것 같습니다!

**Reply by hanseul-lee** [reply]
null과 undefined 관련해 개발자가 명시적으로 입력할 때는 undefined 사용을 지양해야한다고 알고 있습니다.
명시적으로 빈 값에 대해 입력하는 경우는 null을 사용하는 것이 어떠신가요?

> undefined는 개발자가 의도적으로 할당한 값이 아니라 자바스크립트 엔진에 의해 초기화된 값이다. 그렇다면 개발자가 의도적으로 undefined를 할당해야하는 경우가 있을까? 자바스크립트 엔진이 변수 초기화에 사용하는 이 값을 만약 개발자가 마음대로 할당한다면 undefined의 본래의 취지와 어긋날 뿐더러 혼란을 줄 수 있으므로 권장하지 않는다. 그럼 변수의 값이 없다는 것을 명시하고 싶은 경우 어떻게 하면 좋을까? 그런 경우는 undefined를 할당하는 것이 아니라 null을 할당한다.

cf. 
- https://poiemaweb.com/js-data-type-variable#114-undefined
- https://stackoverflow.com/questions/6604749/what-reason-is-there-to-use-null-instead-of-undefined-in-javascript

**Reply by jin-Pro** [reply]
음! 납득이 되는거 같아요! ㅎㅎ 감사합니다

**Reply by pshdev1030** [reply]
> null과 undefined 관련해 개발자가 명시적으로 입력할 때는 undefined 사용을 지양해야한다고 알고 있습니다. 명시적으로 빈 값에 대해 입력하는 경우는 null을 사용하는 것이 어떠신가요?
> 
> > undefined는 개발자가 의도적으로 할당한 값이 아니라 자바스크립트 엔진에 의해 초기화된 값이다. 그렇다면 개발자가 의도적으로 undefined를 할당해야하는 경우가 있을까? 자바스크립트 엔진이 변수 초기화에 사용하는 이 값을 만약 개발자가 마음대로 할당한다면 undefined의 본래의 취지와 어긋날 뿐더러 혼란을 줄 수 있으므로 권장하지 않는다. 그럼 변수의 값이 없다는 것을 명시하고 싶은 경우 어떻게 하면 좋을까? 그런 경우는 undefined를 할당하는 것이 아니라 null을 할당한다.
> 
> cf.
> 
> * https://poiemaweb.com/js-data-type-variable#114-undefined
> * https://stackoverflow.com/questions/6604749/what-reason-is-there-to-use-null-instead-of-undefined-in-javascript

한슬님이 해주신 말씀도 맞긴 한데 d3의 일부 함수는 값의 타입으로 null을 허용하지 않아 undefined를 사용할 수 밖에 없더라구요!

**Reply by hanseul-lee** [reply]
@pshdev1030  오 몰랐네요 감사합니다..!

**Comment by jeonghye-choi** [comment]
### 6. Material-UI 사용

**Reply by pshdev1030** [reply]
시간적 여유가 없기도 해서 저는 쓰는게 낫다는 생각입니다!

**Reply by pshdev1030** [reply]
> 시간적 여유가 없기도 해서 저는 쓰는게 낫다는 생각입니다!

하지만 반응형을 고려하지 않아도 되면 쓰지 않아도 좋을 것 같습니다/

**Reply by pshdev1030** [reply]
https://github.com/githru/githru-vscode-ext/discussions/8#discussioncomment-3405992

추가로 chakra ui가 있고, blueprint가 데스트탑 어플리케이션을 만드는 용도로 반응형을 고려하지 않는걸로 알고있습니다..!

**Reply by jejecrunch** [reply]
MUI를 써도 러닝커브가 그렇게 크지 않다면 쓰는 것도 괜찮다고 생각합니다 ! @pshdev1030 님 말씀대로 시간적 여유가 없기도 하고 저 같은 경우는 MUI를 써본 경험은 있는데 익숙하게 쓸 수는 없어서 다른 분들은 어떨지 모르겠네요 ... 사용하는 건 그렇게 어렵지 않았는데 타입스크립트에서는 써본 경험이 없어서 ㅜㅜ MUI 아니더라도 Ant Design같은 경우는 경험이 많긴 하지만 타입스크립트에서는 처음이라 애매하네요 ㅜㅜ

**Comment by vgihan** [comment]
### 7. 네이밍 컨벤션

네이밍 컨벤션도 정하고 가면 좋을 것 같아서 discussion에 추가하겠습니다 ! 맞추고 가면 좋을 것 같은 네이밍 컨벤션 reply 로 달아주시거나 edit으로 편집해주세요 !

- 컴포넌트 네이밍
- 이벤트 핸들러 함수 네이밍
- boolean 타입 변수 네이밍
- css 네이밍 컨벤션 
- 기타 등등

**Reply by jejecrunch** [reply]
type의 네이밍은 어떻게 하면 좋을까요? 저는 우선 그 객체가 정말 무엇을 나타내는지 명사를 적었는데 다른 분들은 어떤지 궁금하네요.

**Reply by hanseul-lee** [reply]
컴포넌트 머지 후 css 네이밍 컨벤션 관련해 스타일이 중복 적용되는 경우가 발생하는데 
이와 관련해 컨벤션을 정했으면 좋겠습니다.

- ex) BEM
  - summary__name
  - author-bar-chart__name

**Reply by jejecrunch** [reply]
> 컴포넌트 머지 후 css 네이밍 컨벤션 관련해 스타일이 중복 적용되는 경우가 발생하는데 이와 관련해 컨벤션을 정했으면 좋겠습니다.
> 
> * ex) BEM
>   
>   * summary__name
>   * author-bar-chart__name

이렇게 가는 거 좋은 거 같습니다. 생각해보니까 데이터명이 겹치는 곳이 많은데 css 선택자 명을 어떻게 할지 결정하지 않았었네요.

**Reply by hanseul-lee** [reply]
넵! 기한, 영진, 지혜님 회의에서 위 컨벤션에 맞춰가기로 결정하였습니다. 
추후 다른 좋은 의견 있으시면 다시 반영해 보겠습니다.
